app/src/main/kotlin/edu/stanford/bdh/engagehf/questionnaire/spezi/QuestionnaireComposable.kt

```diff
@@ -0,0 +1,142 @@
+package edu.stanford.bdh.engagehf.questionnaire.spezi
+
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.material3.AlertDialog
+import androidx.compose.material3.Text
+import androidx.compose.material3.TextButton
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.ui.Modifier
+import androidx.core.os.bundleOf
+import androidx.fragment.app.setFragmentResultListener
+import androidx.fragment.compose.AndroidFragment
+import androidx.fragment.compose.rememberFragmentState
+import ca.uhn.fhir.context.FhirContext
+import com.google.android.fhir.datacapture.QuestionnaireFragment
+import edu.stanford.spezi.core.design.component.StringResource
+import edu.stanford.spezi.core.utils.extensions.testIdentifier
+import kotlinx.coroutines.launch
+import org.hl7.fhir.r4.model.Questionnaire
+
+@Composable
+fun QuestionnaireComposable(
+    questionnaire: Questionnaire,
+    modifier: Modifier = Modifier,
+    completionStepMessage: String? = null,
+    cancelBehavior: CancelBehavior = CancelBehavior.ShouldConfirmCancel,
+    onResult: suspend (QuestionnaireResult) -> Unit,
+) {
+    val completionItem = remember(completionStepMessage) {
+        completionStepMessage?.let {
+            Questionnaire.QuestionnaireItemComponent().apply {
+                linkId = "completion-step"
+                type = Questionnaire.QuestionnaireItemType.DISPLAY
+                text = completionStepMessage
+            }
+        }
+    }
+    val fullQuestionnaire = remember(questionnaire, completionItem) {
+        completionItem?.let {
+            questionnaire.copy().apply { addItem(it) }
+        } ?: questionnaire
+    }
+    val questionnaireJson = remember(fullQuestionnaire) {
+        FhirContext.forR4()
```

Comment:
encoding can fail, I would avoid crashing at this level, and use:

```kotlin
val questionnaireJson = remember(fullQuestionnaire) {
    runCatching {
        FhirContext.forR4()
            .newJsonParser()
            .encodeResourceToString(fullQuestionnaire)
    }.getOrDefault("")
}
```

We can delegate the invalid json error screen to the library